// netlify-onegraph-internal/dist/nextjs.test.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var fs_1 = require("fs");
var graphql_1 = require("graphql");
var index_1 = require("./index");
var test = function () {
    var sourceGraphQLFilename = "./tests/assets/netlifyGraphOperationsLibrary.graphql";
    var schemaGraphQLFilename = "./tests/assets/netlifyGraphSchema.graphql";
    var sourceGraphQLFile = (0, fs_1.readFileSync)(sourceGraphQLFilename, "utf8");
    var schemaGraphQLFile = (0, fs_1.readFileSync)(schemaGraphQLFilename, "utf8");
    var schema = (0, graphql_1.buildASTSchema)((0, graphql_1.parse)(schemaGraphQLFile));
    var netlifyGraphConfig = {
        netlifyGraphPath: ["..", "..", "lib", "netlifyGraph"],
        webhookBasePath: "/api",
        framework: "Next.js",
        functionsPath: ["pages", "api"],
        graphQLOperationsSourceFilename: [
            "..",
            "..",
            "lib",
            "netlifyGraphOperationsLibrary.graphql",
        ],
        graphQLSchemaFilename: ["..", "..", "lib", "netlifyGraphSchema.graphql"],
        netlifyGraphImplementationFilename: ["..", "..", "lib", "index.js"],
        netlifyGraphTypeDefinitionsFilename: ["..", "..", "lib", "index.d.ts"],
        netlifyGraphRequirePath: ["..", "..", "lib", "netlifyGraph"],
        extension: "ts",
        moduleType: "esm",
        language: "typescript",
        runtimeTargetEnv: "node",
    };
    console.log("config: ", netlifyGraphConfig);
    var result = index_1.NetlifyGraph.generateHandlerSource({
        handlerOptions: {
            postHttpMethod: true,
            useClientAuth: true,
        },
        netlifyGraphConfig: netlifyGraphConfig,
        operationId: "12b5bdea-9bab-4164-a731-5e697b1552be",
        operationsDoc: sourceGraphQLFile,
        schema: schema,
    });
    if (typeof result === "undefined") {
        console.error("No generated next.js code");
    }
    // @ts-ignore
    var exportedFiles = result.exportedFiles;
    console.log("exportedFiles", exportedFiles);
    exportedFiles === null || exportedFiles === void 0 ? void 0 : exportedFiles.forEach(function (element) {
        var _a;
        console.log((_a = element.name) === null || _a === void 0 ? void 0 : _a.join("/"), element.content);
    });
};
test();
