// netlify-onegraph-internal/dist/sanity.test.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var fs_1 = require("fs");
var graphql_1 = require("graphql");
var path = require("path/posix");
var index_1 = require("./index");
var test = function () {
    var sourceGraphQLFilename = "./tests/assets/netlifyGraphOperationsLibrary.graphql";
    var schemaGraphQLFilename = "./tests/assets/netlifyGraphSchema.graphql";
    var sourceGraphQLFile = (0, fs_1.readFileSync)(sourceGraphQLFilename, "utf8");
    var schemaGraphQLFile = (0, fs_1.readFileSync)(schemaGraphQLFilename, "utf8");
    var schema = (0, graphql_1.buildASTSchema)((0, graphql_1.parse)(schemaGraphQLFile));
    var parsedDoc = (0, graphql_1.parse)(sourceGraphQLFile);
    var functions = index_1.NetlifyGraph.extractFunctionsFromOperationDoc(parsedDoc);
    var netlifyGraphConfig = {
        netlifyGraphPath: ["..", "..", "lib", "netlifyGraph"],
        framework: "Next.js",
        webhookBasePath: "/api",
        functionsPath: ["pages", "api"],
        graphQLOperationsSourceFilename: [
            "..",
            "..",
            "lib",
            "netlifyGraphOperationsLibrary.graphql",
        ],
        graphQLSchemaFilename: ["..", "..", "lib", "netlifyGraphSchema.graphql"],
        netlifyGraphImplementationFilename: ["..", "..", "lib", "index.js"],
        netlifyGraphTypeDefinitionsFilename: ["..", "..", "lib", "index.d.ts"],
        netlifyGraphRequirePath: ["..", "..", "lib", "netlifyGraph"],
        extension: "ts",
        moduleType: "esm",
        language: "typescript",
        runtimeTargetEnv: "node",
    };
    var result = index_1.NetlifyGraph.generateHandlerSource({
        handlerOptions: {
            postHttpMethod: true,
            useClientAuth: true,
        },
        netlifyGraphConfig: netlifyGraphConfig,
        operationId: "2b0c3674-06b0-4a84-b296-afa92c10dc6b",
        operationsDoc: sourceGraphQLFile,
        schema: schema,
    });
    if (!result) {
        throw new Error("result is undefined");
    }
    var exportedFiles = result.exportedFiles;
    exportedFiles === null || exportedFiles === void 0 ? void 0 : exportedFiles.forEach(function (exportedFile) {
        var filename = exportedFile.kind === "NamedExportedFile"
            ? path.join.apply(path, exportedFile.name) : "default";
        console.log(filename + ":\n", exportedFile.content);
        (0, fs_1.writeFileSync)("/tmp/" + filename + ".ts", exportedFile.content);
    });
};
test();
