// verdaccio/build/lib/logger/levels.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.calculateLevel = calculateLevel;
exports.subSystemLevels = exports.levelsColors = void 0;

var _kleur = require("kleur");

function calculateLevel(levelCode) {
  switch (true) {
    case levelCode === 10:
      return 'trace';

    case levelCode === 20:
      return 'debug';

    case levelCode === 25:
      return 'http';

    case levelCode === 30:
      return 'info';

    case levelCode === 40:
      return 'warn';

    case levelCode === 50:
      return 'error';

    case levelCode === 60:
      return 'fatal';

    default:
      return 'fatal';
  }
}

const levelsColors = {
  fatal: _kleur.red,
  error: _kleur.red,
  warn: _kleur.yellow,
  http: _kleur.magenta,
  info: _kleur.cyan,
  debug: _kleur.green,
  trace: _kleur.white
};
exports.levelsColors = levelsColors;
var ARROWS;

(function (ARROWS) {
  ARROWS["LEFT"] = "<--";
  ARROWS["RIGHT"] = "-->";
  ARROWS["EQUAL"] = "-=-";
  ARROWS["NEUTRAL"] = "---";
})(ARROWS || (ARROWS = {}));

const subSystemLevels = {
  color: {
    in: (0, _kleur.green)(ARROWS.LEFT),
    out: (0, _kleur.yellow)(ARROWS.RIGHT),
    fs: (0, _kleur.black)(ARROWS.EQUAL),
    default: (0, _kleur.blue)(ARROWS.NEUTRAL)
  },
  white: {
    in: ARROWS.LEFT,
    out: ARROWS.RIGHT,
    fs: ARROWS.EQUAL,
    default: ARROWS.NEUTRAL
  }
};
exports.subSystemLevels = subSystemLevels;
